# SimSnappNet
by Charles-Elie Rabier, Vincent Berry, Jean-Christophe Glaszmann, Fabio Pardi, Scornavacca Celine 


              ISE-M, Univ. Montpellier, CNRS, EPHE, IRD, Montpellier, France

              LIRMM, Univ.Montpellier, CNRS, Montpellier, France

              IMAG, Univ. Montpellier, CNRS, Montpellier, France

              AGAP, CIRAD, Montpellier, France 
************************************************************************************************************************************
Our C++ simulator "SimSnappNet" is build on the software "SimSnapp" written by David Bryant and Remco Bouckaert.

D. Bryant, R. Bouckaert, J. Felsenstein, N.A. Rosenberg, A. RoyChoudhury : Inferring Species Trees Directly from Biallelic Genetic Markers: Bypassing gene trees in a full coalescent analysis. (Molecular Biology and Evolution, Volume 29, Issue 8, 2012) 

***********************************************************************************************************************************

 SimSnappNet generates SNPs data according to a Markovian process that evolved along gene tree branches. The gene tree is random and is obtained according to the Multispecies Network Coalescent (i.e. a coalescent process that evolves inside a phylogenetic network). 
Although SNPS data are generated, SimSnappNet returns only count data ....

This count data can be analyzed with SnappNetForSimSnappNet (see the repository on github). 

***********************************************************************************************************************************

 SimSnappNet is described at  https://rabier.github.io/doc/SimSnappNet.html
 
